src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/IconUrlToImageCacheConverter.cs

```diff
@@ -24,38 +29,127 @@ internal class IconUrlToImageCacheConverter : IValueConverter
+        /// This leads to a memory savings.
+        /// </remarks>
+        /// <param name="values">
+        /// <list type="bullet">
+        /// <item>
+        /// <description>values[0]: IconUrl that points to a URL o a local file</description>
+        /// </item>
+        /// <item>
+        /// <description>values[1]: An <c>PackageArchiveReader</c> to read from the local package for embedded icons</description>
+        /// </item>
+        /// </list>
+        /// </param>
+        /// <param name="targetType">unused</param>
+        /// <param name="parameter">A BitmapImage with the default package icon</param>
+        /// <param name="culture">unused</param>
+        /// <returns>A BitmapSource with the image</returns>
+        public object Convert(object[] values, Type targetType, object parameter, CultureInfo culture)
         {
-            var iconUrl = value as Uri;
+            if (values == null || values.Length == 0)
+            {
+                return null;
+            }
+
+            var iconUrl = values[0] as Uri;
             var defaultPackageIcon = parameter as BitmapSource;
             if (iconUrl == null)
             {
                 return null;
             }
 
-            var cachedBitmapImage = _bitmapImageCache.Get(iconUrl.ToString()) as BitmapSource;
+            var cachedBitmapImage = BitmapImageCache.Get(iconUrl.ToString()) as BitmapSource;
             if (cachedBitmapImage != null)
             {
                 return cachedBitmapImage;
             }
 
             // Some people run on networks with internal NuGet feeds, but no access to the package images on the internet.
             // This is meant to detect that kind of case, and stop spamming the network, so the app remains responsive.
-            if (_errorFloodGate.IsOpen)
+            if (ErrorFloodGate.IsOpen)
             {
                 return defaultPackageIcon;
             }
 
             var iconBitmapImage = new BitmapImage();
             iconBitmapImage.BeginInit();
-            iconBitmapImage.UriSource = iconUrl;
 
+            BitmapSource imageResult;
+
+            // Check if the URI is an Embedded Icon URI
+            if (IsEmbeddedIconUri(iconUrl))
+            {
+                // Check if we have enough info to read the icon from the package
+                if (values.Length == 2 && values[1] is Lazy<PackageReaderBase>)
+                {
+                    try
+                    {
+                        var lazyPar = (Lazy<PackageReaderBase>)values[1];
+                        var par = lazyPar.Value as PackageArchiveReader;
+                        var iconEntry = Uri.UnescapeDataString(iconUrl.Fragment).Substring(1); // skip the '#' in a URI fragment
+                        iconBitmapImage.StreamSource = par.GetEntry(iconEntry).Open(); // This stream is closed in BitmapImage events
```

Comment:
This means that the package file (on disk) will be locked until this stream is disposed.  That means the package file cannot be modified or deleted during this time.

It's not clear to me when the object will be disposed.  As I understand it, WPF maintains a dependency object graph and as objects become unreferenced they become eligible for disposal.  That's at some vague time in the future, and until then the package file could be locked.

Suppose you open PM UI and the list populates with icons from files on disk.  How long are those package files locked?  Does closing PM UI immediately release the locks?

If you can confirm that the locking happens, please open an issue for triaging.

Reply:
> This means that the package file (on disk) will be locked until this stream is disposed. That means the package file cannot be modified or deleted during this time.

- The image stream is closed once the BitmapImage reads the whole image.
- The reader is disposed on the BitmapImage events

> It's not clear to me when the object will be disposed. As I understand it, WPF maintains a dependency object graph and as objects become unreferenced they become eligible for disposal. That's at some vague time in the future, and until then the package file could be locked.

I need to take a look at the docs.

> Suppose you open PM UI and the list populates with icons from files on disk. How long are those package files locked? Does closing PM UI immediately release the locks?

The readers and streams for getting the image are disposed/closed immediately after the icon is read, either on success or failure cases.

> If you can confirm that the locking happens, please open an issue for triaging.

Sure, I will open an issue for investigating further.